Config.ini refactoring

tjk584 opened the pull request:
Reconfigured config.ini to be more user friendly and readable. Moved N200s to an array with individual parameters. The array is then parsed in driveroptions.cpp to produce the same strings as before. TX and RX channels are also now implicitly created based on the N200s specified. 

Associated trello card: https://trello.com/c/OA5bPsH0/35-refactor-config-file

tjk584 commented:
New config format is available here: https://github.com/SuperDARNCanada/borealis_config_files/blob/config_refactoring/sas_config.ini

kevinkrieger commented:
In the file linked here: https://github.com/SuperDARNCanada/borealis_config_files/blob/config_refactoring/sas_config.ini
The interferometer antennas are 1,2,3,4 - should they be 0,1,2,3?

tjk584 commented:
Added logic to calculate main and interferometer array counts from the N200 list in all the options files, removing the need for count parameters in the config file.

tjk584 commented:
When testing, came across two configurations that don't work due to bugs elsewhere in the code (beyond the config file). These configurations are:

1. N200 connected to only interferometer antenna, not main antenna
    - Throws error `free(): invalid size` within usrp driver
    - This problem is likely caused by the device string holding 16 N200 addresses, but the transmit string only having 15 addresses
2. Only TX or RX channels connected for a single N200, not both
    - This is likely caused by main_antenna_count covering both cases, and there not being a distinct count for transmit and receive cases.

kevinkrieger commented on utils/driver_options/set_affinity_options.py:
```diff
@@ -41,7 +41,24 @@ def __init__(self):
+                devices_map[device_num] = n200["addr"]
+
+        addr_idx = 0
+        for device_num in sorted(devices_map.keys()):
```
I think its more pythonic and future-proof to use the python built-in enumerate(...) here instead of keeping track of your own addr_idx

kevinkrieger commented on utils/driver_options/driveroptions.cpp:
```diff
@@ -6,20 +6,110 @@
+                int_antenna_map[int_antenna_num] = device_num;
+                interferometer_antenna_count_++;
+            }
+            if (tx || rx) {
```
I think this if statement (tx || rx) can be moved out of the outer if statement scope

kevinkrieger reviewed (approved):
looks good. Only a few suggestions

RemingtonRohel commented on utils/driver_options/driveroptions.cpp:
```diff
@@ -6,20 +6,110 @@
+        auto addr = iter->second.data();
+
+        // Get rx, tx, and rx_int flags
+        iter++;
```
This block of code tells me that the order of the fields within an n200 field in the config.ini matters. Is this clarified in the documentation somewhere? It might be nice to have it NOT matter, so that the fields can be in whichever order and this code can parse them correctly regardless.

RemingtonRohel reviewed (commented):
It looks like you've hardcoded in the order of fields within an N200 in the config file (ip address, tx, rx, rx_int, etc.). I'm assuming things don't work as expected if two fields are swapped around? I just wonder whether it might be better to set things up such that this is possible, to potentially save some headache for ourselves or others in the future when making new config files.

Also, you swapped some of the fields in the sas_config.ini file of `borealis_config_files`. This reflects what is currently running on the SAS radar?

RemingtonRohel commented on docs/source/config_options.rst:
```diff
@@ -1,194 +1,222 @@
++--------------------------------+-------------------------------+---------------------------------------+
+| main_antennas                  | 0,1,2,3,4,5,6,7,8,9,          | Mapping of main rx/tx channels to     |
+|                                | 10,11,12,13,14,15             | antennas in the main array            |
++--------------------------------+-------------------------------+---------------------------------------+
```
I think these two fields can go now, right? main_antennas and interferometer_antennas? I know these are included in the main branch docs, but I think they are redundant here now with these changes.

tjk584 replied:
I believe they can go also, I just have to add parsing code to create these fields within the various options files to match the code in main. 

It looks like in main, `main_antenna_count` just refers to the number of antennas in the physical array, while in this PR I have it referring to the total number of transmitting/receiving antennas. Should I modify this PR to be more similar to main? Or leave that for merge conflict resolution?

RemingtonRohel replied:
Is it the total number of tx/rx antennas, or the range of the indices? If you are using only antennas [0, 1, 4, 6], will `main_antenna_count` report 4 or 7? 

Honestly I'm not sure, I think we're going to have a heck of a time with the merge conflicts when merging develop into main regardless of what is done here. It might be better just to kick this can down the road rather than relive the same sort of headache twice in trying to figure it out now plus again when we merge that PR.

RemingtonRohel replied:
Nevermind, I think I have a solid opinion on this. I think it would be best to keep `main_antenna_count` and `interferometer_antenna_count` as fields in the config file SEPARATE from the N200 fields. They will be fields that are basically only used to check the validity of the antenna indices within the N200 fields. 

I was looking at how you build up the `..._antenna_count` fields now: [Line 44](https://github.com/SuperDARNCanada/borealis/pull/334/files#diff-148a6e57c058ff1bde9db54347b543dcbf4889310b8ff83240fca83df35363c1R44)
and I think that this can lead to errors down the road if someone makes a mistake in their config file, based on a future check you make: [Line 101](https://github.com/SuperDARNCanada/borealis/pull/334/files#diff-148a6e57c058ff1bde9db54347b543dcbf4889310b8ff83240fca83df35363c1R101). So I would recommend we make the `..._antenna_count` field a constant as I explained in the previous paragraph.

tjk584 commented:
Added code to parse the config file and produce `main_antennas` and `interferometer_antennas` attributes created in https://github.com/SuperDARNCanada/borealis/pull/298. This should make it a little easier to deal with the merge conflicts when merging develop into main.

Removed parsing for `main_antenna_count` and `interferometer_antenna_count`, these parameters will be hard coded into the config file instead and represent the number of physical antennas in each array, regardless of the status of the N200s.
